presto-jdbc/src/main/java/com/facebook/presto/jdbc/PrestoDriver.java

```diff
@@ -57,30 +66,196 @@
+     *       increments queryExecutor.refcount under readlock
+     *       releases read lock
+     *    Caller of close() obtains write lock and:
+     *       decrements refcount
+     *       refcount != 0
+     *       releases write lock
+     *
+     * 4) Simultaneous close() and get(), close() obtains write lock first
+     *    Caller of close() obtains write lock and:
+     *       decrements refcount
+     *       refcount == 0
+     *       invalidates cache, as in 2 above, closing HTTP client.
+     *       releases write lock
+     *    Caller of get() obtains read lock and:
+     *       calls cache.get() which returns a new queryExecutor under the
+     *             cache's own lock
+     *       increments queryExecutor.refcount under readlock
+     *       releases read lock
+     *
+     * The last case of interest is the process that is using PrestoDriver
+     * calling prestoDriver.close(). This invalidates all of the cached
+     * QueryExecutors regardless of their refcount, and in doing so closes
+     * their HTTP clients. If anybody is still trying to uses a
+     * PrestoConnection created by the driver, they're SOL.
+     */
     public PrestoDriver()
     {
-        this.queryExecutor = QueryExecutor.create(DRIVER_NAME + "/" + DRIVER_VERSION);
+        this.cacheLock = new ReentrantReadWriteLock();
+
+        Consumer<QueryExecutor> queryExecutorCloser = new Consumer<QueryExecutor>() {
+            @Override
+            public void accept(QueryExecutor queryExecutor)
+            {
+                cacheLock.writeLock().lock();
+                if (queryExecutor.dereference() == 0) {
+                    queryExecutorCache.invalidate(queryExecutor.getClientProperties());
+                }
+                cacheLock.writeLock().unlock();
+            }
+        };
+
+        this.queryExecutorCache = CacheBuilder.newBuilder()
+                .removalListener(new RemovalListener<Properties, QueryExecutor>()
+                        {
+                            @Override
+                            public void onRemoval(RemovalNotification<Properties, QueryExecutor> notification)
+                            {
+                                notification.getValue().closeHttpClient();
+                            }
+                        })
+                .build(new CacheLoader<Properties, QueryExecutor>() {
+                            @Override
+                            public QueryExecutor load(Properties clientProperties)
+                            {
+                                return QueryExecutor.create(
+                                        DRIVER_NAME + "/" + DRIVER_VERSION,
+                                        clientProperties,
+                                        queryExecutorCloser);
+                            }
+                        });
     }
 
     @Override
     public void close()
     {
-        queryExecutor.close();
+        queryExecutorCache.invalidateAll();
```

Comment:
You should track somehow that this driver instance is closed. So, i won't accept any further `connect` request. Before it was ensured by the `QueryExecutor`.